### PR TITLE
Rubberband fix for selecting large area in Songeditor

### PR DESCRIPTION
**Pull request by BaraMGB**
fixes https://github.com/LMMS/lmms/issues/4974

scrolling and zooming while selecting an area works now. But with 180 additions and 70 deletions this is a big bug fix. I don't know if this should be merged so close before releasing 1.2.

### CODE REVIEW

**Comment by PhysSong** [comment]
> I don't know if this should be merged so close before releasing 1.2.

If you want to test more, we can include this in 1.2.1 instead of 1.2.0.

File: src/gui/editors/SongEditor.cpp [hunk 1]
```diff
@@ -293,6 +300,80 @@ void SongEditor::scrolled( int new_pos )
...
+		//the miditime the mouse is hover
+		MidiTime rubberbandMidipos = MidiTime((m_mousePos.x() - m_trackHeadWidth) / pixelsPerTact()
+											  * MidiTime::ticksPerTact()) + m_currentPosition ;
+		//collect all Tcos
+		QVector<selectableObject *> so;
+		QList<selectableObject *> l = findChildren<selectableObject *>();
+
+		for (QList<selectableObject *>::iterator it = l.begin(); it != l.end(); ++it)
+		{
+			so.push_back( *it );
```
Comment (PhysSong) [inline src/gui/editors/SongEditor.cpp]: No need to waste resources here, you can just use the list directly.

File: src/gui/editors/SongEditor.cpp [hunk 2]
```diff
@@ -293,6 +300,80 @@ void SongEditor::scrolled( int new_pos )
...
+		}
+		//are tcos in the rect of selection?
+		for (QVector<selectableObject *>::iterator it = so.begin(); it != so.end(); ++it)
```
Comment (PhysSong) [inline src/gui/editors/SongEditor.cpp]: It's a good chance to switch to the range-based loop.

File: src/gui/editors/SongEditor.cpp [hunk 3]
```diff
@@ -293,6 +300,80 @@ void SongEditor::scrolled( int new_pos )
...
+		rubberBand()->setGeometry(QRect( origin,
+										 contentWidget()->mapFromParent(QPoint(m_mousePos.x(), m_mousePos.y()))
+										).normalized());
+
+		//the index of the TrackView the mouse is hover
+		const TrackView * tv = trackViewAt(m_mousePos.y() - m_timeLine->height());
+		int rubberBandTrackview = 0;
+		if (tv)
+		{
+			for (QList<TrackView *>::const_iterator it = trackViews().begin(); it != trackViews().end(); ++it)
```
Comment (PhysSong) [inline src/gui/editors/SongEditor.cpp]: A range-based loop will be better.

**Review by PhysSong — COMMENTED** [review]
This is the first part of my review. I will review this PR again in a few days.

**Comment by BaraMGB** [comment]
Thank you for your time.

File: src/gui/editors/SongEditor.cpp [hunk 4]
```diff
@@ -293,6 +300,75 @@ void SongEditor::scrolled( int new_pos )
...
+										 contentWidget()->mapFromParent(QPoint(m_mousePos.x(), m_mousePos.y()))
+										).normalized());
+
+		//the index of the TrackView the mouse is hover
+		const TrackView * tv = trackViewAt(m_mousePos.y() - m_timeLine->height());
+		int rubberBandTrackview = 0;
+		if (tv)
+		{
+			for (auto it : trackViews())
+			{
```
Comment (Reflexe) [inline src/gui/editors/SongEditor.cpp]: Can't understand the reason behind this loop; why won't you use `indexOf`?

File: src/gui/editors/SongEditor.cpp [hunk 5]
```diff
@@ -293,6 +300,75 @@ void SongEditor::scrolled( int new_pos )
...
+											  * MidiTime::ticksPerTact()) + m_currentPosition ;
+		//collect all Tcos
+		auto l = findChildren<selectableObject *>();
+
+		//are tcos in the rect of selection?
+		for (auto &it : l)
+		{
+			TrackContentObjectView * tco = dynamic_cast<TrackContentObjectView*>(it);
+
+			if (trackViews().indexOf(tco->getTrackView())     >= qMin(m_rubberBandStartTrackview, rubberBandTrackview)
```
Comment (Reflexe) [inline src/gui/editors/SongEditor.cpp]: What's the meaning of this statement? `X >= Y` and then `X <= Y`? Am I missing something? 
Also, check for `nullptr` after `dynamic_cast`.

File: src/gui/editors/SongEditor.cpp [hunk 6]
```diff
@@ -404,7 +480,68 @@ void SongEditor::closeEvent( QCloseEvent * ce )
...
+		m_currentZoomingValue = zoomingModel()->value();
+		//paint the rubberband
+		rubberBand()->setEnabled(true);
+		rubberBand()->setGeometry(QRect( m_origin, QSize()));
+		rubberBand()->show();
+
+		//the trackView(index) and the miditime where the mouse has clicked
+		const TrackView * tv = trackViewAt(_me->pos().y() - m_timeLine->height());
+		m_rubberBandStartTrackview = 0;
+		if (tv)
```
Comment (Reflexe) [inline src/gui/editors/SongEditor.cpp]: A duplicate of line 350. Please extract it into something like `updateRubberbandSelectionForMousePos(pos)`.

**Review by Reflexe — CHANGES_REQUESTED** [review]

File: src/gui/editors/SongEditor.cpp (hunk 5 above)
Comment (BaraMGB) [reply src/gui/editors/SongEditor.cpp]: You can make the selection from top to bottom or from bottom to top. The right side of the if-statement is the exact opposite in the second if-statement. Is it in the first statement `m_rubberBandStartTrackview` so it is mandatory in the second if-statement `rubberbandTrackView` and vice versa. So you can test whether the selection is made from below or from above.

But maybe I missed something? It gave me a real headache. Perhaps there is a simpler way?

File: include/Track.h [hunk 7]
```diff
@@ -210,6 +210,12 @@ class TrackContentObjectView : public selectableObject, public ModelView
 	{
 		return m_tco;
 	}
+
```
[the remark this answers is not in the document]
Comment (BaraMGB) [reply include/Track.h]: It was protected but I need this outside of the class. Can I avoid this move?

File: src/gui/editors/SongEditor.cpp (hunk 4 above)
Comment (Reflexe) [reply src/gui/editors/SongEditor.cpp]: I understand. We could add `indexOfTrackView(const TrackView)` with `std::distance(std::find(m_trackViews.begin(), m_trackViews.end(), tv), m_trackViews.begin()` instead.

File: src/gui/editors/SongEditor.cpp (hunk 5 above)
Comment (Reflexe) [reply src/gui/editors/SongEditor.cpp]: I don't understand: the condition `(X >= Y && X <= Y)` will be true only if `X==Y`.

Comment (Reflexe) [reply src/gui/editors/SongEditor.cpp]: My bad; missed something.

**Review by Reflexe — APPROVED** [review]

**Comment by BaraMGB** [comment]
@Reflexe Thanks a lot for reviewing this.
@PhysSong Do you have any objection about this PR?

File: src/gui/editors/SongEditor.cpp [hunk 8]
```diff
@@ -239,6 +243,9 @@ SongEditor::SongEditor( Song * song ) :
 					this, SLOT( scrolled( int ) ) );
 	connect( m_song, SIGNAL( lengthChanged( int ) ),
 			this, SLOT( updateScrollBar( int ) ) );
+	connect( m_leftRightScroll, SIGNAL(valueChanged(int)),this, SLOT(updateRubberband()));
+	connect( contentWidget()->verticalScrollBar(), SIGNAL(valueChanged(int)),this, SLOT(updateRubberband()));
```
Comment (PhysSong) [inline src/gui/editors/SongEditor.cpp]: There's whitespace after the first opening parenthesis, but there isn't one before the last closing parenthesis. It looks a bit inconsistent to me.

File: src/gui/editors/SongEditor.cpp [hunk 9]
```diff
@@ -293,6 +300,57 @@ void SongEditor::scrolled( int new_pos )
...
+
+		//take care of the scrollbar position
+		int hs = (m_leftRightScroll->value() - m_scrollPos.x()) * pixelsPerTact();
+		int vs = contentWidget()->verticalScrollBar()->value() - m_scrollPos.y();
+
+		//the adjusted origin point
+		QPoint origin = QPoint(qMax(originX - hs, m_trackHeadWidth), m_origin.y() - vs);
+
+		//paint the rubber band rect
+		rubberBand()->setGeometry(QRect( origin,
```
Comment (PhysSong) [inline src/gui/editors/SongEditor.cpp]: Similar here.

File: src/gui/editors/SongEditor.cpp [hunk 10]
```diff
@@ -293,6 +300,57 @@ void SongEditor::scrolled( int new_pos )
...
+
+		//the miditime the mouse is hover
+		MidiTime rubberbandMidipos = MidiTime((m_mousePos.x() - m_trackHeadWidth)
+											  / static_cast<int>(pixelsPerTact())
+											  * MidiTime::ticksPerTact()) + m_currentPosition ;
+		//are tcos in the rect of selection?
+		for (auto &it : findChildren<selectableObject *>())
+		{
+			TrackContentObjectView * tco = dynamic_cast<TrackContentObjectView*>(it);
+			if(tco)
```
Comment (PhysSong) [inline src/gui/editors/SongEditor.cpp]:
```suggestion
			if (tco)
```

File: src/gui/editors/SongEditor.cpp [hunk 11]
```diff
@@ -404,7 +462,55 @@ void SongEditor::closeEvent( QCloseEvent * ce )
...
+	{
+		//we save the position of both scrollbars into a QPoint
+		m_scrollPos = QPoint(m_leftRightScroll->value(), contentWidget()->verticalScrollBar()->value());
+		//and the mouse position
+		m_origin = contentWidget()->mapFromParent(QPoint(_me->pos().x(), _me->pos().y()));
+		//and the zooming
+		m_currentZoomingValue = zoomingModel()->value();
+		//paint the rubberband
+		rubberBand()->setEnabled(true);
+		rubberBand()->setGeometry(QRect( m_origin, QSize()));
```
Comment (PhysSong) [inline src/gui/editors/SongEditor.cpp]: Here as well.

File: src/gui/editors/SongEditor.cpp [hunk 12]
```diff
@@ -404,7 +462,55 @@ void SongEditor::closeEvent( QCloseEvent * ce )
...
+		rubberBand()->show();
+
+		//the trackView(index) and the miditime where the mouse has clicked
+		m_rubberBandStartTrackview = trackIndexFromSelectionPoint(_me->y());
+
+		m_rubberbandStartMidipos = MidiTime((_me->x() - m_trackHeadWidth)
+											/ pixelsPerTact() * MidiTime::ticksPerTact())
+											+ m_currentPosition;
+	}
+	QWidget::mousePressEvent( _me );
```
Comment (PhysSong) [inline src/gui/editors/SongEditor.cpp]: Removing whitespaces in parentheses in the new code will be better.

File: src/gui/editors/SongEditor.cpp [hunk 13]
```diff
@@ -404,7 +462,55 @@ void SongEditor::closeEvent( QCloseEvent * ce )
...
+}
+
+
+
+
+void SongEditor::mouseReleaseEvent(QMouseEvent *_me)
+{
+	rubberBand()->hide();
+	rubberBand()->setEnabled( false );
+	QWidget::mouseReleaseEvent( _me );
```
Comment (PhysSong) [inline src/gui/editors/SongEditor.cpp]: Here as well.

**Review by PhysSong — COMMENTED** [review]
Here are minor comments regarding whitespaces. I'm testing the functionality and will let you know if there are any issues or not.

**Comment by PhysSong** [comment]
I found that lassoing with Shift+drag on the timeline is broken. @BaraMGB Could you look into it?

**Comment by BaraMGB** [comment]
@PhysSong Thank you. I'll look into it. :+1:

File: src/gui/editors/SongEditor.cpp [hunk 14]
```diff
@@ -80,15 +79,21 @@ SongEditor::SongEditor( Song * song ) :
 	m_zoomingModel(new ComboBoxModel()),
 	m_scrollBack( false ),
 	m_smoothScroll( ConfigManager::inst()->value( "ui", "smoothscroll" ).toInt() ),
-	m_mode(DrawMode)
+	m_mode(DrawMode),
+	m_origin(),
+	m_scrollPos(),
+	m_mousePos(),
+	m_rubberBandStartTrackview(),
+	m_rubberbandStartMidipos(),
```
Comment (PhysSong) [inline src/gui/editors/SongEditor.cpp]: I think initializing `int` values with explicit 0s will be better.

File: src/gui/editors/SongEditor.cpp [hunk 15]
```diff
@@ -293,6 +302,98 @@ void SongEditor::scrolled( int new_pos )
...
+
+		//we save the position of scrollbars, mouse position and zooming level
+		m_origin = QPoint(xStart, 0);
+		m_scrollPos = QPoint(m_leftRightScroll->value(), contentWidget()->verticalScrollBar()->value());
+		m_currentZoomingValue = zoomingModel()->value();
+
+		//calculate the song position where the mouse was clicked
+		m_rubberbandStartMidipos = MidiTime((xStart - m_trackHeadWidth)
+											/ pixelsPerTact() * MidiTime::ticksPerTact())
+											+ m_currentPosition;
```
Comment (PhysSong) [inline src/gui/editors/SongEditor.cpp]: You should set `m_rubberBandStartTrackview` to 0 as well. If I select some TCOs in track views and then from the timeline, some TCOs are not selected although the rubberband displays correctly.

Comment (BaraMGB) [reply src/gui/editors/SongEditor.cpp]: fixed it.

**Comment by BaraMGB** [comment]
@PhysSong can this be merged or do you want to test it more?

File: include/SongEditor.h [hunk 16]
```diff
@@ -85,6 +88,9 @@ public slots:
 
 protected:
 	virtual void closeEvent( QCloseEvent * ce );
+	virtual void mousePressEvent( QMouseEvent * _me );
+	virtual void mouseMoveEvent( QMouseEvent * _me );
+	virtual void mouseReleaseEvent( QMouseEvent * _me );
```
Comment (PhysSong) [inline include/SongEditor.h]: No underscores anymore...
```
• Function parameters are not prefixed with '_' anymore.
```
https://github.com/LMMS/lmms/wiki/Coding-conventions

File: src/gui/editors/SongEditor.cpp [hunk 17]
```diff
@@ -404,7 +506,53 @@ void SongEditor::closeEvent( QCloseEvent * ce )
 		hide();
 	}
 	ce->ignore();
- }
+}
+
+
+
+
+void SongEditor::mousePressEvent(QMouseEvent *_me)
```
Comment (PhysSong) [inline src/gui/editors/SongEditor.cpp]: The same here and below.

**Comment by PhysSong** [comment]
The functionality looks okay, but I think we have some time to test this more because this PR targets the stable branch.

**Comment by BaraMGB** [comment]
@PhysSong Do you think, this should go into 1.2.1?

**Comment by PhysSong** [comment]
I think it's okay to merge this now. As far as I tested, this won't bring any serious regressions.

**Pull request merged**